### PR TITLE
Move types.h #undefs for wincrypt.h compatibility

### LINKED ISSUE #9981: Header file conflict code bug in windows
```c++
//reappear bug
#include <openssl/sha.h>
#include <openssl/aes.h>
#include <openssl/evp.h>
#ifndef CPPHTTPLIB_OPENSSL_SUPPORT
#define CPPHTTPLIB_OPENSSL_SUPPORT
#endif
#include <httplib.h>
```

the conflict between the above code and the [httplib project](https://github.com/yhirose/cpp-httplib)
[reappear bug example](https://github.com/HeMinzhang/opensslbug/blob/master/src/makeBug/main.cpp)

[openssl github issue](https://github.com/openssl/openssl/issues/9981)

[httplib github issue](https://github.com/yhirose/cpp-httplib/issues/220)

[Previous discussion in qt bugreport](https://bugreports.qt.io/browse/QTBUG-78642)

### PR DESCRIPTION
**Pull request by samuel-lee-msft**
+ Always undef the symbols that may have been #define-d
  by wincrypt.h after the first inclusion of types.h to
  avoid errors from wincrypt.h symbols being used to
  compile OpenSSL code
+ Also need to remove #pragma once for this approach to work
+ Define WINCRYPT_USE_SYMBOL_PREFIX to enable wincrypt
  symbol prefix at some point in future

Fixes #9981

- [x] tests are added or updated

### CODE REVIEW

File: include/openssl/types.h
Comment (samuel-lee-msft) [inline include/openssl/types.h:76]: In prototyping the modified wincrypt.h, I could not find any reference to `PKCS7_ISSUER_AND_SERIAL` in Windows headers - so I removed the undef in this PR.

File: include/openssl/types.h
Comment (davidben) [inline include/openssl/types.h:15]: Oh interesting. Is this something the Windows headers will later consume to emit different constants?

Comment (samuel-lee-msft) [reply include/openssl/types.h:15]: Yes - that is the plan

Comment (samuel-lee-msft) [reply include/openssl/types.h:15]: I've got this approved by folks on my team to update wincrypt.h to respect WINCRYPT_USE_SYMBOL_PREFIX in future.

**Comment by samuel-lee-msft** [comment]
@mspncp any chance of a review on this? Thanks for your feedback on the issue

**Comment by mspncp** [comment]
Sorry @samuel-lee-msft, I was a little bit distracted in the last two weeks by covid and didn't notice that you had posted a replacement for #18079. I'll take a look in the next few days.

**Comment by h-vetinari** [comment]
Line endings in `test/build.info` changed spuriously, leading to 100% diff.

**Comment by samuel-lee-msft** [comment]
> Line endings in `test/build.info` changed spuriously, leading to 100% diff.

Good catch - fixed in latest

**Review by mspncp — COMMENTED** [review]
Sorry for letting you wait so long. Looks good now. Just one question:

File: include/openssl/types.h
Comment (mspncp) [reply include/openssl/types.h:12]: Ok, I got it know. Lines 10-21 need to be parsed always. It seems like my brain is still affected by post-covid symptoms.

**Review by mspncp — APPROVED** [review]

File: include/openssl/types.h
Comment (samuel-lee-msft) [reply include/openssl/types.h:12]: Ah - the point is to move the #undefs to outside of the guarded region, so on WIN32, the undefs will be performed unconditionally (i.e. specifically if types.h has been included before).

This is so we can support a situation where we have the sequence of includes:
```
#include <openssl/types.h>
#include <wincrypt.h>
#include <openssl/types.h>  // <-- need to undef the wincrypt.h #defines at this point. With pragma once this include does nothing
```

**Comment by mspncp** [comment]
ping @openssl/committers for second review

**Comment by mspncp** [comment]
> ping @t8m - any steps I need to take to get this completed?

No steps required except for pinging the reviewers once in a while 😁

**Review by t8m — APPROVED** [review]

**Comment by mspncp** [comment]
Merged to master as 3c58d447497b37f7b4f458aaa2956a7e226c6d65. Thany you very much for your contribution and your patience!

File: include/openssl/types.h
Comment (davidben) [reply include/openssl/types.h:15]: @samuel-lee-msft Do you have an estimate for when the updated wincrypt.h will be available?

**Pull request closed without merging**
